### PR TITLE
Libexpat 2.2.2+

### PR DESCRIPTION
**Pull request by catalinr**
Updated to commit `5f1f1f1d98fef39ef885d4271029b4c3fd814a2d`

### CODE REVIEW

**Comment by vadz** [comment]
Thanks a lot to you (and @hartwork) for working on this!

**Comment by lanurmi** [comment]
Unfortunately building master on CentOS 6 (and possibly others) is broken since this was merged.

`In file included from ../src/expat/lib/xmlparse.c:38:
../src/expat/lib/siphash.h:353:5: warning: "SIPHASH_MAIN" is not defined
../src/expat/lib/xmlparse.c:58:3: error: #error You do not have support for any sources of high quality entropy enabled. For end user security, that is probably not what you want. Your options include: * Linux + glibc >=2.25 (getrandom): HAVE_GETRANDOM, * Linux + glibc <2.25 (syscall SYS_getrandom): HAVE_SYSCALL_GETRANDOM, * BSD / macOS (arc4random_buf): HAVE_ARC4RANDOM_BUF, * libbsd (arc4random_buf): HAVE_ARC4RANDOM_BUF + HAVE_LIBBSD, * Windows (RtlGenRandom): _WIN32. If insist on not using any of these, bypass this error by defining XML_POOR_ENTROPY; you have been warned. For CMake, one way to pass the define is: cmake -DCMAKE_C_FLAGS="-pipe -O2 -DHAVE_SYSCALL_GETRANDOM" . If you have reasons to patch this detection code away or need changes to the build system, please open a bug. Thank you!
make: *** [wxexpat_xmlparse.o] Error 1
`

**Comment by catalinr** [comment]
There have been several changes for this to libexpat. Hopefully it was fixed in newer commits. I'll try to find out a bit later. Otherwise if you have a minute you can probably check libexpat commits too and see if anyone in particular sounds relevant.

**Comment by hartwork** [comment]
I was expecting that wxWidgets only uses bundled Expat on non-Unix operating systems that do not come with Expat, i.e. not Linux. How far off am I?
@lanurmi Since CentOS 6 comes with Expat, did you force wxWidgets to build with bundled Expat some way?

**Comment by catalinr** [comment]
@hartwork actually wxWidgets always uses built-in Expat.

**Comment by hartwork** [comment]
Can we change the default to `--with-expat=sys` on Linux? Linux distros package wxWTK for security already, it would fix compilation for @lanurmi and would reduce the number of outdated vulnerable binary Expat in the world for the future. Do you expect problems from `--with-expat=sys` for a default?

**Comment by catalinr** [comment]
I was a bit hasty. It looks like it is trying to get system Expat by default, and switches to built-in only upon failing to find it or when forced by an explicit option. So the question about forcing it is valid.

**Comment by hartwork** [comment]
I see!
@lanurmi could you check if you have text `system expat library not found, will use built-in instead` in the output of configure and also check if you have Expat headers installed (e.g. `/usr/include/expat.h`)? If not, please install the regular Expat development package for CentOS and try again.
If that still doesn't work, probably a new ticket would be good place to continue this discussion.

**Comment by lanurmi** [comment]
I was indeed configuring just with `../configure` without any flags, so I did not force anything one way or another. I did not have the `expat-devel` package installed, which is why the built-in is used. Installing `expat-devel` and re-running configure + make solves the problem, but this is a new undocumented dependency for compiling wx.

**Comment by vadz** [comment]
There can be good reasons for using built-in Expat even under non-MSW systems, for example I always use it when building macOS applications distributed in binary form. And while I don't do this personally, I think people distributing Linux binaries do it too, for the same reasons.

So I'm pretty sure we're going to have problems due to this. What I don't understand is why didn't Expat configure detect the appropriate method? I.e. why do we have this problem in the first place?

**Comment by hartwork** [comment]
Hi!

> There can be good reasons for using built-in Expat even under non-MSW systems, for example I always use it when building macOS applications distributed in binary form. And while I don't do this personally, I think people distributing Linux binaries do it too, for the same reasons.

macOS and Linux distros come with Expat out of the box and their security teams are keeping Expat up to date with security fixes. Looking at [wxWidget's use of Expat](https://github.com/wxWidgets/wxWidgets/blob/9af43c8c06482af77468af2daff66289632c0d24/src/xml/xml.cpp#L895) I don't see need for using a non-system copy of Expat that is likely slower to receive security updates, if at all. [Bundles are a security issue in general](https://wiki.gentoo.org/wiki/Why_not_bundle_dependencies#Security_implications), so it would be cool if they would only be used when _really_ needed. That applies to Windows only probably.

> So I'm pretty sure we're going to have problems due to this. What I don't understand is why didn't Expat configure detect the appropriate method? I.e. why do we have this problem in the first place?

`configure` doesn't detect any of the supported methods, because CentOS 6 doesn't have any: It's too old. The `#error` check was meant to guard against accidental lack of any of these due to misconfiguration with build systems other than GNU Autoconf, originally. That's why configure doesn't error about it here. Let me think about that more.

**Comment by vadz** [comment]
System libs are used by default if available, and there are options to both force or disable using them. However I think you underestimate how ancient some systems which still need to be targeted today are, I definitely can't drop CentOS/RHEL 6 support yet and I'm willing to bet that the version of Expat bundled with wx is much newer than the one available in it. I'm also almost sure there are people who still have to support CentOS 5 too...

Also, this is not a great argument, but the vast majority of wx users won't ever use Expat with untrusted input at all as it's used for parsing resource files. So in many cases it is more important to have some easily available version of Expat, even if it's not the latest one, than to not have it at all because without it the application just can't be compiled/used.

To summarize, we really need to do something to allow building Expat included in wx even under old systems such as CentOS 6.

**Comment by hartwork** [comment]
Hi Vadim,

I don't agree on that assessment but I'm in with getting configure-based wxWidgets Expat free of compilation errors on CentOS 6 again. Expat Git `master` by now has support for reading `/dev/urandom` directly and should not show that error reported above, as of a the latest commit just now. For someone with a CentosOS 6 (or Ubuntu trusty or Debian wheezy) at hand, please run:
```console
git clone --depth 1 https://github.com/libexpat/libexpat.git
cd libexpat/expat
./buildconf.sh && ./configure && make
```

**Comment by lanurmi** [comment]
> For someone with a CentosOS 6 (or Ubuntu trusty or Debian jessie) at hand, please run:

It successfully compiles without errors on C6.

**Comment by vadz** [comment]
Thanks @hartwork for the change, I don't necessarily agree with this assessment myself, but I just know we'd be getting flak from many people if we broke this.

@catalinr I guess it would be better to update to latest Expat master again instead of just cherry-picking this change?

**Comment by catalinr** [comment]
I guess so. But if I don't get to do it tomorrow it will probably have to wait at least until the next weekend.

**Comment by hartwork** [comment]
Which has the upside that I get a chance to get a new release out the door until then.

**Comment by hartwork** [comment]
Expat 2.2.3 has just been released, includes fix for DLL hijacking on Windows and the CentOS compile fixes that you tested earlier.
